pkg/registry/registry.go

```diff
@@ -0,0 +1,264 @@
+	User     string
+	Password string
+}
+
+type Repository struct {
+	Name string
+	Tags []string
+}
+
+func (r *Registry) Login() error {
+	fmt.Printf("Logging in destination container registry...\n")
+
+	args := []string{"login", r.Name, fmt.Sprintf("-u%s", r.Credentials.User), fmt.Sprintf("-p%s", r.Credentials.Password)}
+
+	err := executeCmd(dockerBinaryName, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) Logout() error {
+	fmt.Printf("Logging out of destination container registry...\n")
+
+	var args []string
+
+	if r.Name == "" {
+		args = []string{"logout"}
+	} else {
+		args = []string{"logout", r.Name}
+	}
+
+	err := executeCmd(dockerBinaryName, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) ListRepositoryTags(repo string) ([]string, error) {
+	fmt.Printf("\nReading list of tags from source registry for %#q repository...\n", repo)
+
+	endpoint := fmt.Sprintf("%s/v2/%s/tags/list", r.Address, repo)
+
+	type tagsJSON struct {
+		Tags []string `json:"tags"`
+	}
+
+	var tagsData tagsJSON
+
+	var tags []string
+	{
+		nextEndpoint := fmt.Sprintf("%s", endpoint)
+		for {
+			req, err := http.NewRequest("GET", nextEndpoint, nil)
+			if err != nil {
+				return []string{}, microerror.Mask(err)
+			}
+
+			resp, err := r.HttpClient.Do(req)
+			if err != nil {
+				return []string{}, microerror.Mask(err)
+			}
+
+			defer resp.Body.Close()
+			body, err := ioutil.ReadAll(resp.Body)
+			if err != nil {
+				return []string{}, microerror.Mask(err)
+			}
+
+			err = json.Unmarshal(body, &tagsData)
+			if err != nil {
+				return []string{}, microerror.Mask(err)
+			}
+
+			for _, tag := range tagsData.Tags {
+				if len(tag) < tagLengthLimit {
+					tags = append(tags, tag)
+				}
+			}
+
+			linkHeader := resp.Header.Get("Link")
+			if linkHeader != "" {
+				nextEndpoint = fmt.Sprintf("%s%s", r.Address, getLink(linkHeader))
+			} else {
+				break
+			}
+		}
+	}
+
+	return tags, nil
+
+}
+
+func (r *Registry) PullImage(repo, tag string) error {
+	image := fmt.Sprintf("%s/%s:%s", r.Name, repo, tag)
+
+	args := []string{"pull", image}
+
+	err := executeCmd(dockerBinaryName, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) PushImage(repo, tag string) error {
+	image := fmt.Sprintf("%s/%s:%s", r.Name, repo, tag)
+
+	args := []string{"push", image}
+
+	err := executeCmd(dockerBinaryName, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (r *Registry) RemoveImage(repo, tag string) error {
+	image := fmt.Sprintf("%s/%s:%s", r.Name, repo, tag)
+
+	args := []string{"rmi", image}
+
+	err := executeCmd(dockerBinaryName, args)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func RetagImage(repo, tag, srcRegistry, dstRegistry string) error {
+	var srcImage, dstImage string
+
+	if srcRegistry == "" {
+		srcImage = fmt.Sprintf("%s:%s", repo, tag)
```

Comment:
We don't need to support short names with "docker.io" right, or we still do?

Reply:
no, leftover from the time when I didn't know I can actually use `docker.io` registry name